etcd/watchmanager/manager.go

```diff
@@ -0,0 +1,148 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package watchmanager
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3x/log"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/uber-go/tally"
+)
+
+// NewWatchManager creates a new watch manager
+func NewWatchManager(opts Options) (WatchManager, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	scope := opts.InstrumentsOptions().MetricsScope()
+	return &manager{
+		opts:   opts,
+		logger: opts.InstrumentsOptions().Logger(),
+		m: metrics{
+			etcdWatchCreate: scope.Counter("etcd-watch-create"),
+			etcdWatchError:  scope.Counter("etcd-watch-error"),
+			etcdWatchReset:  scope.Counter("etcd-watch-reset"),
+		},
+		updateFn:      opts.UpdateFn(),
+		tickAndStopFn: opts.TickAndStopFn(),
+	}, nil
+}
+
+type manager struct {
+	opts   Options
+	logger xlog.Logger
+	m      metrics
+
+	updateFn      UpdateFn
+	tickAndStopFn TickAndStopFn
+}
+
+type metrics struct {
+	etcdWatchCreate tally.Counter
+	etcdWatchError  tally.Counter
+	etcdWatchReset  tally.Counter
+}
+
+func (w *manager) watchChanWithTimeout(key string) (clientv3.WatchChan, error) {
+	doneCh := make(chan struct{})
+
+	var watchChan clientv3.WatchChan
+	go func() {
+		watchChan = w.opts.Watcher().Watch(
+			clientv3.WithRequireLeader(context.Background()),
+			key,
+			w.opts.WatchOptions()...,
+		)
+		close(doneCh)
+	}()
+
+	select {
+	case <-doneCh:
+		return watchChan, nil
+	case <-time.After(w.opts.WatchChanInitTimeout()):
+		return nil, fmt.Errorf("etcd watch chan creation timed-out on key: %s", key)
+	}
+}
+
+func (w *manager) Watch(key string) {
+	ticker := time.Tick(w.opts.WatchChanCheckInterval())
+
+	var (
+		watchChan clientv3.WatchChan
+		err       error
+	)
+	for {
+		if watchChan == nil {
+			w.m.etcdWatchCreate.Inc(1)
+			watchChan, err = w.watchChanWithTimeout(key)
+			if err != nil {
+				w.logger.Errorf("could not create etcd watch, %v", err)
+
+				// NB(cw) when we failed to create a etcd watch channel
+				// we do a get for now and will try to recreate the watch chan later
+				if err = w.updateFn(key); err != nil {
```

Comment:
So if we can never get a watch established, we can still (regularly) try to get the latest value and push to the users rather than sorely depend on the etcd watch for notifications